tests/quickcheck.rs

```diff
@@ -0,0 +1,305 @@
+    gt_lt_eq
+        .iter()
+        .fold(0, |acc, e| if *e { acc + 1 } else { acc })
+        == 1
+}
+
+#[quickcheck]
+fn quickcheck_unsigned_sub_primitive(a: u128, b: u128) -> bool {
+    if b < a {
+        BigUint::from(a - b) == BigUint::from(a) - BigUint::from(b)
+    } else {
+        BigUint::from(b - a) == BigUint::from(b) - BigUint::from(a)
+    }
+}
+
+#[quickcheck]
+fn quickcheck_signed_sub_primitive(a: i128, b: i128) -> bool {
+    if b < a {
+        BigInt::from(a - b) == BigInt::from(a) - BigInt::from(b)
+    } else {
+        BigInt::from(b - a) == BigInt::from(b) - BigInt::from(a)
+    }
+}
+
+#[quickcheck]
+/// Tests correctness of subtraction assuming addition is correct
+fn quickcheck_unsigned_sub(a: BigUint, b: BigUint) -> bool {
+    if b < a {
+        a.clone() - b.clone() + b == a
+    } else {
+        b.clone() - a.clone() + a == b
+    }
+}
+
+#[quickcheck]
+/// Tests correctness of subtraction assuming addition is correct
+fn quickcheck_signed_sub(a: BigInt, b: BigInt) -> bool {
+    if b < a {
+        a.clone() - b.clone() + b == a
+    } else {
+        b.clone() - a.clone() + a == b
+    }
+}
+
+#[quickcheck]
+fn quickcheck_unsigned_div_primitive(a: u128, b: u128) -> TestResult {
+    if b == 0 {
+        TestResult::discard()
+    } else {
+        TestResult::from_bool(BigUint::from(a / b) == BigUint::from(a) / BigUint::from(b))
+    }
+}
+
+#[quickcheck]
+fn quickcheck_signed_div_primitive(a: i128, b: i128) -> TestResult {
+    if b == 0 {
+        TestResult::discard()
+    } else {
+        TestResult::from_bool(BigInt::from(a / b) == BigInt::from(a) / BigInt::from(b))
+    }
+}
+
+#[quickcheck]
+fn quickcheck_unsigned_pow_zero(a: BigUint) -> bool {
+    a.pow(0_u32) == BigUint::one()
+}
+
+#[quickcheck]
+fn quickcheck_unsigned_pow_one(a: BigUint) -> bool {
+    a.pow(1_u32) == a
+}
+
+#[quickcheck]
+fn quickcheck_unsigned_sqrt(a: BigUint) -> bool {
+    (a.clone() * a.clone()).sqrt() == a
+}
+
+#[quickcheck]
+fn quickcheck_unsigned_cbrt(a: BigUint) -> bool {
+    (a.clone() * a.clone() * a.clone()).cbrt() == a
+}
+
+#[quickcheck]
+fn quickcheck_signed_cbrt(a: BigInt) -> bool {
+    (a.clone() * a.clone() * a.clone()).cbrt() == a
+}
+
+#[quickcheck]
+fn quickcheck_unsigned_conversion(a: BigUint, radix: u8) -> TestResult {
+    let radix = radix as u32;
+    if radix > 36 || radix < 2 {
+        return TestResult::discard();
+    }
+    let string = a.to_str_radix(radix);
+    TestResult::from_bool(a == BigUint::from_str_radix(&string, radix).unwrap())
+}
+
+#[quickcheck]
+fn quickcheck_signed_conversion(a: BigInt, radix: u8) -> TestResult {
+    let radix = radix as u32;
+    if radix > 36 || radix < 2 {
+        return TestResult::discard();
+    }
+    let string = a.to_str_radix(radix);
+    TestResult::from_bool(a == BigInt::from_str_radix(&string, radix).unwrap())
+}
+
+#[test]
+fn quicktest_shift() {
+    let gen = StdThreadGen::new(usize::max_value());
+    let mut qc = QuickCheck::with_gen(gen);
```

Comment:
I think it might be a good idea to do every test like this - to have a custom generator that generates full ranges for all types that aren't bigger than `usize`. Shame we still can't get the full range for `u128` this way, since `StdThreadGen::new` takes a `usize`.

Reply:
I tried to set an environment variable - `export QUICKCHECK_GENERATOR_SIZE=4294967295`, but this variable controls both the length of vectors generated and the maximum value of integers generated (so does `QuickCheck::with_gen()`), so setting it so high gives an allocation error `memory allocation of 16691263072 bytes failederror: process didn't exit successfully:` `...`.

 We could write our own code to determine the length of a vector generated with Arbitrary, but use the default Arbitrary for u32. This as an alternative to using the custom `StdThreadGen` in tests.

I really wish quickcheck's defaults were a better for our use case.

Reply:
I'm going to call this good for now. I ended up making a custom `StdThreadGen` for all the tests based on primitives i.e.
```rust
let gen = StdThreadGen::new(usize::max_value());
let mut qc = QuickCheck::with_gen(gen);
```
quickchecks taking `BigUint`/`BigInt` use the standard quickcheck gen (which isn't ideal, but I expect to find many bugs). This gen implementation can only check values where all `BigDigit`s have values 0-100 and there are 0-100 big digits. good enough.